stan/math/rev/core/var.hpp

```diff
@@ -177,12 +149,284 @@ class var_value {
+   * then (a *= b) behaves exactly the same way as (a = a * b).
+   * Note that the result is an assignable lvalue.
+   *
+   * @param b The variable to multiply this variable by.
+   * @return The result of multiplying this variable by the
+   * specified variable.
+   */
+  inline var_value<T>& operator*=(const var_value<T>& b);
+
+  /**
+   * The compound multiply/assignment operator for scalars (C++).
+   *
+   * If this variable is a and the argument is the scalar b, then
+   * (a *= b) behaves exactly the same way as (a = a * b).  Note
+   * that the result is an assignable lvalue.
+   *
+   * @param b The scalar to multiply this variable by.
+   * @return The result of multiplying this variable by the specified
+   * variable.
+   */
+  inline var_value<T>& operator*=(T b);
+
+  /**
+   * The compound divide/assignment operator for variables (C++).  If this
+   * variable is a and the argument is the variable b, then (a /= b)
+   * behaves exactly the same way as (a = a / b).  Note that the
+   * result is an assignable lvalue.
+   *
+   * @param b The variable to divide this variable by.
+   * @return The result of dividing this variable by the
+   * specified variable.
+   */
+  inline var_value<T>& operator/=(const var_value<T>& b);
+
+  /**
+   * The compound divide/assignment operator for scalars (C++).
+   *
+   * If this variable is a and the argument is the scalar b, then
+   * (a /= b) behaves exactly the same way as (a = a / b).  Note
+   * that the result is an assignable lvalue.
+   *
+   * @param b The scalar to divide this variable by.
+   * @return The result of dividing this variable by the specified
+   * variable.
+   */
+  inline var_value<T>& operator/=(T b);
+
+  /**
+   * Write the value of this autodiff variable and its adjoint to
+   * the specified output stream.
+   *
+   * @param os Output stream to which to write.
+   * @param v Variable to write.
+   * @return Reference to the specified output stream.
+   */
+  friend std::ostream& operator<<(std::ostream& os, const var_value<T>& v) {
+    if (v.vi_ == nullptr) {
+      return os << "uninitialized";
+    }
+    return os << v.val();
+  }
+
+  inline auto& eval() noexcept { return *this; }
+};
+
+/**
+ * Independent (input) and dependent (output) variables for gradients.
+ *
+ * This class acts as a smart pointer, with resources managed by
+ * an arena-based memory manager scoped to a single gradient
+ * calculation.
+ *
+ * A var is constructed with a type `T` and used like any
+ * other scalar. Arithmetical functions like negation, addition,
+ * and subtraction, as well as a range of mathematical functions
+ * like exponentiation and powers are overridden to operate on
+ * var values objects.
+ * @tparam T An Floating point type.
+ */
+template <typename T>
+class var_value<
+    T, require_t<bool_constant<is_eigen<T>::value || !is_plain_type<T>::value
+                               || is_matrix_cl<T>::value>>> {
```

Comment:
Because some Eigen types like the return from `rowwise()` don't actually inherit from `EigenBase`. Should we have some other detection like `is_eigen_expr`?

Reply:
Also I need the rowwise thing so we can efficiently do `reverse()` when users ask for `min_max_index` subsetting like `A[4:1,]`

Reply:
Yeah `!is_plain_type<T>::value` is a tiny bit confusing. It is fine if you want to leave it as it is. I don't think adding `is_eigen_expr` is necessary (and if it would not be a negation of `is_plain_type` it would really get confusing). Instead I would suggest adding `is_eigen_vectorwise` that would detect these rowwise and colwise expressions.

Reply:
Aight yeah so with the `rowwise_reverse()` stuff we can just move this back to `is_eigen<T> || is_matrix_cl<T>` since those return objects the inherit from `DenseBase`